subsys/bluetooth/audio/has_client.c

```diff
@@ -895,108 +1413,191 @@ int bt_has_client_presets_read(struct bt_has *has, uint8_t start_index, uint8_t
 		return -EOPNOTSUPP;
 	}
 
-	if (atomic_test_bit(inst->flags, HAS_DISCOVER_IN_PROGRESS) ||
-	    atomic_test_and_set_bit(inst->flags, HAS_CP_OPERATION_IN_PROGRESS)) {
-		return -EBUSY;
-	}
-
 	opcode = sync ? BT_HAS_OP_SET_ACTIVE_PRESET_SYNC : BT_HAS_OP_SET_ACTIVE_PRESET;
 
 	return preset_set(inst, opcode, index);
 }
 
-int bt_has_client_preset_next(struct bt_has *has, bool sync)
+int bt_has_client_cmd_preset_next(struct bt_has_client *inst, bool sync)
 {
-	struct has_inst *inst = HAS_INST(has);
 	uint8_t opcode;
 
-	LOG_DBG("conn %p sync %d", (void *)inst->conn, sync);
+	LOG_DBG("inst %p sync %d", inst, sync);
 
-	if (!inst->conn) {
+	CHECKIF(inst == NULL) {
 		return -ENOTCONN;
 	}
 
-	if (sync && !bt_has_feat_is_preset_oob_sync_supported(inst->has.features)) {
-		return -EOPNOTSUPP;
+	if (inst->state != STATE_CONNECTED) {
+		return -ENOTCONN;
 	}
 
-	if (atomic_test_bit(inst->flags, HAS_DISCOVER_IN_PROGRESS) ||
-	    atomic_test_and_set_bit(inst->flags, HAS_CP_OPERATION_IN_PROGRESS)) {
-		return -EBUSY;
+	if (sync && !bt_has_features_check_preset_sync_supp(inst->features)) {
+		return -EOPNOTSUPP;
 	}
 
 	opcode = sync ? BT_HAS_OP_SET_NEXT_PRESET_SYNC : BT_HAS_OP_SET_NEXT_PRESET;
 
 	return preset_set_next_or_prev(inst, opcode);
 }
 
-int bt_has_client_preset_prev(struct bt_has *has, bool sync)
+int bt_has_client_cmd_preset_prev(struct bt_has_client *inst, bool sync)
 {
-	struct has_inst *inst = HAS_INST(has);
 	uint8_t opcode;
 
-	LOG_DBG("conn %p sync %d", (void *)inst->conn, sync);
+	LOG_DBG("inst %p sync %d", inst, sync);
 
-	if (!inst->conn) {
+	CHECKIF(inst == NULL) {
 		return -ENOTCONN;
 	}
 
-	if (sync && !bt_has_feat_is_preset_oob_sync_supported(inst->has.features)) {
-		return -EOPNOTSUPP;
+	if (inst->state != STATE_CONNECTED) {
+		return -ENOTCONN;
 	}
 
-	if (atomic_test_bit(inst->flags, HAS_DISCOVER_IN_PROGRESS) ||
-	    atomic_test_and_set_bit(inst->flags, HAS_CP_OPERATION_IN_PROGRESS)) {
-		return -EBUSY;
+	if (sync && !bt_has_features_check_preset_sync_supp(inst->features)) {
+		return -EOPNOTSUPP;
 	}
 
 	opcode = sync ? BT_HAS_OP_SET_PREV_PRESET_SYNC : BT_HAS_OP_SET_PREV_PRESET;
 
 	return preset_set_next_or_prev(inst, opcode);
 }
 
+int bt_has_client_info_get(const struct bt_has_client *inst, struct bt_has_client_info *info)
+{
+	CHECKIF(inst == NULL) {
+		LOG_ERR("NULL inst");
+
+		return -EINVAL;
+	}
+
+	CHECKIF(info == NULL) {
+		LOG_ERR("NULL info");
+
+		return -EINVAL;
+	}
+
+	info->addr = &inst->addr;
+	info->id = inst->id;
+	info->type = bt_has_features_get_type(inst->features);
+	info->features = inst->features;
+	info->caps = get_capabilities(inst);
+	info->active_index = inst->active_index;
+
+	return 0;
+}
+
+int bt_has_client_conn_get(const struct bt_has_client *inst, struct bt_conn **conn)
+{
+	CHECKIF(inst == NULL) {
+		LOG_ERR("NULL inst");
+
+		return -EINVAL;
+	}
+
+	CHECKIF(conn == NULL) {
+		LOG_ERR("NULL conn");
+
+		return -EINVAL;
+	}
```

Comment:
IMO when we return an error code in a function, it doesn't make sense to also log it as an error

Reply:
Agreed, but I would suggest to use `LOG_DBG` in that case